src/SFML/Window/Win32/JoystickImpl.cpp

```diff
@@ -33,9 +33,128 @@
+
+        const GUID GUID_POV           = {0xa36d02f2, 0xc9f3, 0x11cf, {0xbf, 0xc7, 0x44, 0x45, 0x53, 0x54, 0x00, 0x00}};
+
+        const GUID GUID_RxAxis        = {0xa36d02f4, 0xc9f3, 0x11cf, {0xbf, 0xc7, 0x44, 0x45, 0x53, 0x54, 0x00, 0x00}};
+        const GUID GUID_RyAxis        = {0xa36d02f5, 0xc9f3, 0x11cf, {0xbf, 0xc7, 0x44, 0x45, 0x53, 0x54, 0x00, 0x00}};
+    }
+
+    struct ObjectDataFormat
+    {
+        ObjectDataFormat()
+        {
+            const DWORD axisType   = DIDFT_AXIS   | DIDFT_OPTIONAL | DIDFT_ANYINSTANCE;
+            const DWORD povType    = DIDFT_POV    | DIDFT_OPTIONAL | DIDFT_ANYINSTANCE;
+            const DWORD buttonType = DIDFT_BUTTON | DIDFT_OPTIONAL | DIDFT_ANYINSTANCE;
+
+            data[0].pguid = &guids::GUID_XAxis;
+            data[0].dwOfs = DIJOFS_X;
+
+            data[1].pguid = &guids::GUID_YAxis;
+            data[1].dwOfs = DIJOFS_Y;
+
+            data[2].pguid = &guids::GUID_ZAxis;
+            data[2].dwOfs = DIJOFS_Z;
+
+            data[3].pguid = &guids::GUID_RxAxis;
+            data[3].dwOfs = DIJOFS_RX;
+
+            data[4].pguid = &guids::GUID_RyAxis;
+            data[4].dwOfs = DIJOFS_RY;
+
+            data[5].pguid = &guids::GUID_RzAxis;
+            data[5].dwOfs = DIJOFS_RZ;
+
+            data[6].pguid = &guids::GUID_Slider;
+            data[6].dwOfs = DIJOFS_SLIDER(0);
+
+            data[7].pguid = &guids::GUID_Slider;
+            data[7].dwOfs = DIJOFS_SLIDER(1);
+
+            for (int i = 0; i < 8; ++i)
+            {
+                data[i].dwType = axisType;
+                data[i].dwFlags = DIDOI_ASPECTPOSITION;
+            }
+
+            for (int i = 0; i < 4; ++i)
+            {
+                data[8 + i].pguid = &guids::GUID_POV;
+                data[8 + i].dwOfs = static_cast<DWORD>(DIJOFS_POV(i));
+                data[8 + i].dwType = povType;
+                data[8 + i].dwFlags = 0;
+            }
+
+            for (int i = 0; i < sf::Joystick::ButtonCount; ++i)
+            {
+                data[8 + 4 + i].pguid = NULL;
+                data[8 + 4 + i].dwOfs = static_cast<DWORD>(DIJOFS_BUTTON(i));
+                data[8 + 4 + i].dwType = buttonType;
+                data[8 + 4 + i].dwFlags = 0;
+            }
+
+            format.dwSize = sizeof(DIDATAFORMAT);
+            format.dwObjSize = sizeof(DIOBJECTDATAFORMAT);
+            format.dwFlags = DIDFT_ABSAXIS;
+            format.dwDataSize = sizeof(DIJOYSTATE);
+            format.dwNumObjs = 8 + 4 + sf::Joystick::ButtonCount;
+            format.rgodf = data;
+        }
+
+        DIOBJECTDATAFORMAT data[8 + 4 + sf::Joystick::ButtonCount];
+        DIDATAFORMAT format;
+    };
+
+    const ObjectDataFormat objectDataFormat;
+
+    IDirectInput8W* directInput = NULL;
+    HMODULE dinput8dll = NULL;
+    HRESULT(WINAPI *DirectInput8CreateFunc)(HINSTANCE, DWORD, REFIID, LPVOID*, LPUNKNOWN);
```

Comment:
`DirectInput8CreateFunc` doesn't follow our naming conventions (starts with capital letter). There's no need for it to be a global, it is only used in `initializeDInput`.

The DirectInput-related variables could be stuffed into a structure (like `DirectInputContext`), to limit the number of globals and make things clearer.

`objectDataFormat` is also used locally in a single place, but since its address is taken, it may have to live as long as DInput is used; do you know more about this stuff?